Release/src/http/client/http_client_asio.cpp

```diff
@@ -1966,6 +1966,124 @@ void asio_client::send_request(const std::shared_ptr<request_context>& request_c
+    }
+}
+
+static bool is_recognized_redirection(status_code code)
+{
+    // other 3xx status codes are not handled
+    return is_retrieval_redirection(code) || is_unchanged_redirection(code);
+}
+
+static bool is_retrieval_request(method method)
+{
+    return methods::GET == method || methods::HEAD == method;
+}
+
+static const std::vector<utility::string_t> request_body_header_names =
+{
+    header_names::content_encoding,
+    header_names::content_language,
+    header_names::content_length,
+    header_names::content_location,
+    header_names::content_type
+};
+
+pplx::task<http_response> follow_redirect(http_client_config config,
+                                          http_request request,
+                                          http_response response,
+                                          std::vector<uri> followed_urls = {})
+{
+    // Return immediately if the response is not a supported redirection
+    if (!is_recognized_redirection(response.status_code()))
+        return pplx::task_from_result(response);
+
+    // Although not required by RFC 7231, config may limit the number of automatic redirects
+    if (config.max_redirects() <= followed_urls.size())
+        return pplx::task_from_result(response);
+
+    // Can't very well automatically redirect if the server hasn't provided a Location
+    if (!response.headers().has(header_names::location))
+        return pplx::task_from_result(response);
+
+    const uri location(request.absolute_uri().resolve_uri(response.headers()[header_names::location]));
+
+    // Config may prohibit automatic redirects from HTTPS to HTTP
+    if (!config.https_to_http_redirects() && request.absolute_uri().scheme() == _XPLATSTR("https")
+        && location.scheme() != _XPLATSTR("https"))
+        return pplx::task_from_result(response);
+
+    // "A client SHOULD detect and intervene in cyclical redirections."
+    if (followed_urls.end() != std::find(followed_urls.begin(), followed_urls.end(), location))
+        return pplx::task_from_result(response);
+
+    // Implementation limitation: cannot redirect e.g. a POST request using the same method
+    // since the request body may have been consumed
+    if (!is_retrieval_request(request.method()) && !is_retrieval_redirection(response.status_code()))
+        return pplx::task_from_result(response);
+
+    const auto method = is_retrieval_request(request.method()) ? request.method() : methods::GET;
+
+    http_request redirect(method);
+
+    // Generally, sending the same headers as the original request makes sense
+    // However, Content-* headers do not make sense if no body is being sent
+    redirect.headers() = request.headers();
+    for (const auto& content_header : request_body_header_names)
+    {
+        redirect.headers().remove(content_header);
+    }
+
+    http_client client(location, config);
+
+    return client.request(redirect, request._cancellation_token()).then([=](http_response response) mutable
```

Comment:
It would be better to use a struct instead of a lambda here so that the vector can be moved into it rather than copied.

Reply:
Addressed in 6a4830bb